006-graphql-pagination.md

```diff
@@ -0,0 +1,84 @@
+# GraphQL Pagination
+
+We should use [Relay Cursor Connections](https://facebook.github.io/relay/graphql/connections.htm#) to paginate results in GraphQL. 
+
+**Discussion:** https://github.com/DoSomething/rfcs/pull/6
+
+## Problem
+
+We currently return plain lists, like `[Post]`, for paginated GraphQL queries:
+
+```gql
+{
+  posts(count: 10, page: $page) {
+    id
+    type
+    url
+  }
+}
+```
+
+This gives us no way to know whether there's another page of results without making a request to check. Because we built these on top of standard `OFFSET` pagination, there are also some performance issues with larger datasets & the potential for duplicate results to appear (described in more detail under "Alternatives").
+
+## Proposal
+
+We should use [Relay Cursor Connections](https://facebook.github.io/relay/graphql/connections.htm#) to paginate results in GraphQL, and add REST API support to paginate via cursors (e.g. "get me all the results after #1015" rather than "get me the tenth page of results") to Northstar and Rogue.
+
+```gql
+{
+  paginatedPosts(first: 10, after: $cursor) {
+    edges {
+      cursor
+      node {
+        id
+        type
+        url
+      }
+    }
+    pageInfo {
+      hasNextPage
+    }
+  }
+}
+```
+
+This relies on generating "opaque" cursors for each element in the results. The cursor, e.g. `Nzk1MS4yNDA3Nw==` encodes information about where the item appears in the results (the ID & the sorted column value, if applicable).
```

Comment:
Maybe worth linking to where you've already implemented this in graphql & rogue for a concrete example?

Reply:
Yes! This is implemented on Rogue's campaigns & posts endpoints (since we need to be able to adjust the database query for pagination) in the [`HasCursor` trait](https://github.com/DoSomething/rogue/blob/564255e20b15942d9a0c05d7fa1d169c50e9c60a/app/Models/Traits/HasCursor.php).

For unsorted lists, the cursor is simply the ID (so for example `NA==`, which base-64 decoded is just `4`). So we'd use `cursor[after]=NA==` to get results after ID 4.

For sorted lists, the cursor is the ID & sorted column value (so for example, `OTAyOC41OTA3Nw==` becomes `9028.59077` for campaign ID 9028 with 59,077 pending posts). This allows us to get results after the sorted column value (and handle cases where multiple items have the same sorted column value, like a bunch of campaigns with zero pending posts).

Reply:
_So_ helpful, thanks for the link and explanation!